File: haskell/Problem36.hs
-- Copyright 2015 Mitchell Kember. Subject to the MIT License.
-- Project Euler: Problem 36
-- Double-base palindromes

module Problem36 where

import Common (isPalindrome, digits)

import Data.Bits (unsafeShiftL, unsafeShiftR, (.&.), (.|.))

isBinPalindrome :: Int -> Bool
isBinPalindrome n = odd n && binReverse n == n
  where
    binReverse = go 0
    go r 0 = r
    go r n = go r' n'
      where
        r' = (r `unsafeShiftL` 1) .|. (n .&. 1)
        n' = n `unsafeShiftR` 1

solve :: Int
solve = sum . filter works $ [1..999999]
  where
    works n = isBinPalindrome n && isPalindrome (digits n)

File: c++/src/problem_36.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 36
// Circular primes

#include "common.hpp"

namespace problem_36 {

long solve() {
	long total = 0;
	for (long i = 1; i < 1000000; ++i) {
		if (common::is_palindrome_binary(i) && common::is_palindrome(i)) {
			total += i;
		}
	}
	return total;
}

} // namespace problem_36
